### ISSUE TITLE
external/adapter will preventDefault() on all drop events for external drags if imported

### ISSUE THREAD

**Issue by hitsthings** [atlassian/pragmatic-drag-and-drop#65]
Hi Alex et al! Thanks for open-sourcing this!

In our SPA we have a page using pragmatic-drag-and-drop and another page using raw drag events with Prosemirror.

The existence of pragmatic-drag-and-drop in the SPA, even though it's not used on the current page, causes Prosemirror to ignore our other drop events when starting from external drags. This is because pragmatic will `preventDefault()` on drop events at the `window` level, and Prosemirror won't handle events if `defaultPrevented`. There is no way to work around this that I've found because:

- [external-adapter uses an iife for `startup()`](https://github.com/atlassian/pragmatic-drag-and-drop/blob/main/packages/core/src/adapter/external-adapter.ts#L214)
- [which calls adapter.registerUsage()](https://github.com/atlassian/pragmatic-drag-and-drop/blob/main/packages/core/src/adapter/external-adapter.ts#L223)
- [which calls `register` with `mount`](https://github.com/atlassian/pragmatic-drag-and-drop/blob/main/packages/core/src/make-adapter/make-adapter.ts#L76)
- [which calls another `registerUsage`](https://github.com/atlassian/pragmatic-drag-and-drop/blob/main/packages/core/src/ledger/usage-ledger.ts#L49)
- [which calls `mount`](https://github.com/atlassian/pragmatic-drag-and-drop/blob/main/packages/core/src/ledger/usage-ledger.ts#L37)
- [which sets up a `"dragenter"` handler on the `window`.](https://github.com/atlassian/pragmatic-drag-and-drop/blob/main/packages/core/src/adapter/external-adapter.ts#L77)
- [which will call api.start on every applicable dragenter](https://github.com/atlassian/pragmatic-drag-and-drop/blob/main/packages/core/src/adapter/external-adapter.ts#L131)
- [which calls lifecycle.start](https://github.com/atlassian/pragmatic-drag-and-drop/blob/main/packages/core/src/make-adapter/make-adapter.ts#L59)
- [which sets up a drop handler against window that will immediately call `preventDefault()` on every drop](https://github.com/atlassian/pragmatic-drag-and-drop/blob/main/packages/core/src/ledger/lifecycle-manager.ts#L260)

So ultimately - an iife on import causes all drop events for external drags to have defaultPrevented for the life of the page.

It would be nice if pragmatic only mutated the drop event on elements that are drop targets. The lifecycle.start could be added only on a call to `dropTargetForExternal`, and only against the target element until cleanup runs.

Currently Prosemirror is prevented from handling drops when pragmatic has been loaded due to the defaultPrevented check here:
![image](https://github.com/atlassian/pragmatic-drag-and-drop/assets/631812/ae06f2b7-2ffa-4fbb-ac86-7f7708bffa73)

I can't see a way to intercept any of this, but if you have any workarounds, please let me know! Cheers!

Take care,
Adam

**Comment by alexreardon** [comment]
I think I have a fix 🫡

I'll post an update soon

**Comment by alexreardon** [comment]
https://www.loom.com/share/c98395f1988741339f42d491ccd060a5

**Comment by alexreardon** [comment]
@hitsthings can you please try out this fix using https://www.npmjs.com/package/patch-package

```diff
{
        type: 'drop',
        listener(event: DragEvent) {
          // A "drop" can only happen if the browser allowed the drop

+         /**
+          * If not over any drop targets, then something unmanaged
+          * by pdnd accepted the drop - so we should cancel our operation.
+          * Not calling `event.preventDefault()`
+          */
+          if (!state.current.dropTargets.length) {
+            cancel();
+            return;
+          }
```

**Comment by alexreardon** [comment]
Also, 

- What version of prosemirror are you using?
- Would it be possible for you to provide a sandbox environment so we can ensure everything is working as expected

**Comment by alexreardon** [comment]
@hitsthings i've been working on adding tests for this change and it's looking good. I would love if you could validate the fix in your setup just to make sure that we are all good

**Comment by alexreardon** [comment]
After the proposed external changes, I can see that `eventBelongsToView` is no longer returning `false` in the "drop" event

**Comment by hitsthings** [comment]
Holy shit dude, you went all out on this! Sorry I'm only just coming back to it now, and thanks for the super support! I took a look and with the patch, I can drop images into the editor again. But I think it was already obvious you've completely annihilated this bug 😅 

Likely not relevant to know our version anymore, but we've got a ton of prosemirror stuff going on via TipTap so there are a lot of different versions we're on. prosemirror-view is probably the most relevant and we are on 1.31.4 of it.

I appreciate the thoughtful Loom video too, great discussion of the issues. It all makes perfect sense and I like that you listen for drags even before a drop target is added and the fix to cancel only the drop event makes sense. It was only the preventDefault() that was hurting us, but that's done now.

Thanks a ton, man!
